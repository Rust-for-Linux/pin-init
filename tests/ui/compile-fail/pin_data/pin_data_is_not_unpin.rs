use pin_init::*;

#[pin_data]
struct Foo {
    a: u8,
    #[pin]
    _pin: core::marker::PhantomPinned,
}

fn assert_unpin<T: core::marker::Unpin>() {}

fn main() {
    assert_unpin::<Foo>()
}
