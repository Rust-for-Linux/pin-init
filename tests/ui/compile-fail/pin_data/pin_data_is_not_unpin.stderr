error[E0277]: `PhantomPinned` cannot be unpinned
  --> tests/ui/compile-fail/pin_data/pin_data_is_not_unpin.rs:13:20
   |
13 |     assert_unpin::<Foo>()
   |                    ^^^ within `__Unpin<'_>`, the trait `Unpin` is not implemented for `PhantomPinned`
   |
   = note: consider using the `pin!` macro
           consider using `Box::pin` if you need to access the pinned value outside of the current scope
note: required because it appears within the type `__Unpin<'_>`
  --> tests/ui/compile-fail/pin_data/pin_data_is_not_unpin.rs:3:1
   |
3  | #[pin_data]
   | ^^^^^^^^^^^
note: required for `Foo` to implement `Unpin`
  --> tests/ui/compile-fail/pin_data/pin_data_is_not_unpin.rs:3:1
   |
3  | #[pin_data]
   | ^^^^^^^^^^^ unsatisfied trait bound introduced here
4  | struct Foo {
   |        ^^^
note: required by a bound in `assert_unpin`
  --> tests/ui/compile-fail/pin_data/pin_data_is_not_unpin.rs:10:20
   |
10 | fn assert_unpin<T: core::marker::Unpin>() {}
   |                    ^^^^^^^^^^^^^^^^^^^ required by this bound in `assert_unpin`
   = note: this error originates in the attribute macro `pin_data` (in Nightly builds, run with -Z macro-backtrace for more info)
